compiler/rustc_lint/src/types.rs

```diff
@@ -1394,3 +1396,211 @@ impl<'tcx> LateLintPass<'tcx> for VariantSizeDifferences {
         }
     }
 }
+
+declare_lint! {
+    /// The `invalid_atomic_ordering` lint detects using passing an `Ordering`
+    /// to an atomic operation that does not support that ordering.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// # use core::sync::atomic::{AtomicU8, Ordering};
+    /// let atom = AtomicU8::new(0);
+    /// let value = atom.load(Ordering::Release);
+    /// # let _ = value;
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Some atomic operations are only supported for a subset of the
+    /// `atomic::Ordering` variants. Passing an unsupported variant will cause
+    /// an unconditional panic at runtime, which is detected by this lint.
+    ///
+    /// This lint will trigger in the following cases: (where `AtomicType` is an
+    /// atomic type from `core::sync::atomic`, such as `AtomicBool`,
+    /// `AtomicPtr`, `AtomicUsize`, or any of the other integer atomics).
+    ///
+    /// - Passing `Ordering::Acquire` or `Ordering::AcqRel` to
+    ///   `AtomicType::store`.
+    ///
+    /// - Passing `Ordering::Release` or `Ordering::AcqRel` to
+    ///   `AtomicType::load`.
+    ///
+    /// - Passing `Ordering::Relaxed` to `core::sync::atomic::fence` or
+    ///   `core::sync::atomic::compiler_fence`.
+    ///
+    /// - Passing `Ordering::Release` or `Ordering::AcqRel` as the failure
+    ///   ordering for any of `AtomicType::compare_exchange`,
+    ///   `AtomicType::compare_exchange_weak`, or `AtomicType::fetch_update`.
+    ///
+    /// - Passing in a pair of orderings to `AtomicType::compare_exchange`,
+    ///   `AtomicType::compare_exchange_weak`, or `AtomicType::fetch_update`
+    ///   where the failure ordering is stronger than the success ordering.
+    INVALID_ATOMIC_ORDERING,
+    Deny,
+    "usage of invalid atomic ordering in atomic operations and memory fences"
+}
+
+declare_lint_pass!(InvalidAtomicOrdering => [INVALID_ATOMIC_ORDERING]);
+
+impl InvalidAtomicOrdering {
+    fn type_is_atomic(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+        const ATOMIC_TYPES: &[Symbol] = &[
+            sym::atomic_bool_type,
+            sym::atomic_ptr_type,
+            sym::atomic_usize_type,
+            sym::atomic_u8_type,
+            sym::atomic_u16_type,
+            sym::atomic_u32_type,
+            sym::atomic_u64_type,
+            sym::atomic_u128_type,
+            sym::atomic_isize_type,
+            sym::atomic_i8_type,
+            sym::atomic_i16_type,
+            sym::atomic_i32_type,
+            sym::atomic_i64_type,
+            sym::atomic_i128_type,
+        ];
+        if let ty::Adt(&ty::AdtDef { did, .. }, _) = cx.typeck_results().expr_ty(expr).kind() {
+            ATOMIC_TYPES.iter().any(|sym| cx.tcx.is_diagnostic_item(*sym, did))
```

Comment:
Yes, we should IMO

Reply:
It already exists: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.get_diagnostic_item
It has a warning that it's slower than `is_diagnostic_item` (presumably because it invalids more of the incremental cache). Should I use it anyway?

Reply:
Also, I'm not sure what you're suggesting the new code should look like. Don't I need to check all the types in the array no matter what?

Reply:
> It already exists

That gets a `DefId` from a name. I'm suggesting the opposite.

My second suggestion is to make the parent module a diagnostic item instead of the types and then just check the name of the type.